projects/core/src/occ/services/occ-fields.service.ts

```diff
@@ -0,0 +1,179 @@
+import { Injectable } from '@angular/core';
+import { extractFields, mergeFields, parseFields } from '../utils/occ-fields';
+import { ScopedData } from '../../model/scoped-data';
+import { map, shareReplay } from 'rxjs/operators';
+import { Observable } from 'rxjs';
+import { HttpClient } from '@angular/common/http';
+
+/**
+ * Intermediate model to accommodate all data needed to perform occ fields optimizations
+ * wrapping ScopedData with url and fields
+ */
+export interface OccFieldsModels {
+  /** Url (with fields) to load scoped data */
+  url?: string;
+  /** extracted fields object, used to extract data from broader model */
+  fields?: object;
+  /** scoped data model */
+  model: ScopedData<any>;
+}
+
+/**
+ * Grouped rest calls with optimal urls
+ *
+ * One url groups all scopes it covers with related occFieldsModels
+ */
+export interface OccMergedUrls {
+  [optimalUrl: string]: {
+    [scope: string]: OccFieldsModels;
+  };
+}
+
+/**
+ * Helper service for optimizing endpoint calls to occ backend
+ */
+@Injectable({
+  providedIn: 'root',
+})
+export class OccFieldsService {
+  constructor(protected http: HttpClient) {}
+
+  protected FIELDS_PARAM = 'fields';
+
+  /**
+   * Optimize occ endpoint calls merging requests to the same url by merging field parameters
+   *
+   * @param occFieldsModels
+   * @param dataFactory
+   */
+  optimalLoad<T>(
+    occFieldsModels: OccFieldsModels[],
+    dataFactory?: (url: string) => Observable<T>
+  ): ScopedData<T>[] {
+    const result = [];
+
+    if (!dataFactory) {
+      dataFactory = url => this.http.get<any>(url);
+    }
+
+    const mergedUrls = this.getMergedUrls(occFieldsModels);
+
+    Object.entries(mergedUrls).forEach(
+      ([url, scopes]: [
+        string,
+        {
+          [scope: string]: OccFieldsModels;
+        }
+      ]) => {
+        const scopesForUrl = Object.values(scopes);
+
+        if (scopesForUrl.length === 1) {
+          // only one scope for url
+          result.push({
+            ...scopesForUrl[0].model,
+            data$: dataFactory(url),
+          });
+        } else {
+          // multiple scopes per url
+          const data$ = dataFactory(url).pipe(
+            shareReplay(1),
+            map(data => JSON.parse(JSON.stringify(data)))
+          );
+
+          scopesForUrl.forEach(modelData => {
+            result.push({
+              ...modelData.model,
+              data$: data$.pipe(
+                map(product => extractFields<T>(product, modelData.fields))
+              ),
+            });
+          });
+        }
+      }
+    );
+
+    return result;
+  }
+
+  /**
+   * Merge similar occ endpoints calls by merging fields parameter
+   *
+   * @param models
+   */
+  getMergedUrls<T>(models: OccFieldsModels[]): OccMergedUrls {
```

Comment:
We don't merge URLs here, but group similar URLs and merge their fields

Reply:
👍